### PR TITLE
Implemented Shannon-Expansion

### PR DESCRIPTION
**Pull request by mikand**
Implemented the classical shannon expansion method as a quantifier eliminator for the BOOL logic. This addresses #183.

### CODE REVIEW

File: pysmt/solvers/qelim.py [hunk 1]
```diff
@@ -42,3 +48,39 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+    LOGICS = [pysmt.logics.BOOL]
+
+    def __init__(self, environment, logic=None):
+        IdentityDagWalker.__init__(self, env=environment)
+        self.logic = logic
+
+    def eliminate_quantifiers(self, formula):
+        logic = get_logic(formula, self.env)
+        if not logic <= pysmt.logics.BOOL:
+            raise NotImplementedError("Shannon quantifier elimination only "\
```
Comment (marcogario) [inline pysmt/solvers/qelim.py]: Isn't this too strong?
Shouldn't we check that the quantified variables are Boolean? The expression:
``` Exists x. x & (r < 5)``` becomes ```(False & (r < 5)) | (True & (r < 5))```, no?
Am I missing something?

File: pysmt/utils.py [hunk 2]
```diff
@@ -1,7 +1,12 @@
 import io
+import itertools
 from six import PY2
 from fractions import Fraction
 
+# powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)
```
Comment (marcogario) [inline pysmt/utils.py]: Make this into a proper function.

File: pysmt/solvers/qelim.py [hunk 3]
```diff
@@ -42,3 +48,39 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+                                      "(detected logic is: %s)" % str(logic))
+
+        return self.walk(formula)
+
+    def build_assignments(self, variables):
+        for s in powerset(variables):
+            yield dict((x, self.mgr.Bool(x in s)) for x in variables)
+
+    def walk_forall(self, formula, args, **kwargs):
+        conj = []
```
Comment (marcogario) [inline pysmt/solvers/qelim.py]: Unify the two functions. Return the list of elements:
```
def walk_forall(...):
  return self.mgr.And(self._expand(formula, args))
```

File: pysmt/test/test_shannon_expansion.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    def test_nested(self):
+        f = Exists([self.x], ForAll([self.y], Or(self.x, self.y)))
+        g = qelim(f, solver_name="shannon")
+        g = g.simplify()
+        self.assertEqual(g, TRUE())
+
+
+    def test_examples_solving(self):
+        for example in get_example_formulae():
```
Comment (marcogario) [inline pysmt/test/test_shannon_expansion.py]: We really need to reconsider the interface of ```get_example_formulae()``` I think selection by logic is one of the most common things we do in the tests...

File: pysmt/test/test_shannon_expansion.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            else:
+                self.assertTrue(g.is_false())
+
+            f = ForAll(fv, example.expr)
+            g = qelim(f, solver_name="shannon").simplify()
+            if example.is_valid:
+                self.assertTrue(g.is_true())
+            else:
+                self.assertTrue(g.is_false())
+
```
Comment (marcogario) [inline pysmt/test/test_shannon_expansion.py]: If we extend it for SMT formulae, then we should probably compare the result against a solver (i.e., Z3).

**Comment by marcogario** [comment]
Good!

Open points:
- Can we extend to Theory?
- Squash everything
- Update docs/Changes

File: pysmt/test/test_shannon_expansion.py (hunk 5 above)
Comment (mikand) [inline pysmt/test/test_shannon_expansion.py]: We cannot solve the satisfiability problem if a single non-boolean variable is present in the formula...

File: pysmt/solvers/qelim.py (hunk 3 above)
Comment (mikand) [inline pysmt/solvers/qelim.py]: OK

File: pysmt/utils.py (hunk 2 above)
Comment (mikand) [inline pysmt/utils.py]: OK

File: pysmt/solvers/qelim.py (hunk 1 above)
Comment (mikand) [inline pysmt/solvers/qelim.py]: In principle yes, the problem is that this QE is complete for BOOL, but incomplete for any other theory (we have no way of distinguishing a formula having only boolean variables quantified).

The pysmt framework for QE does not fit well for this case: shall we declare this QE as supporting any logic and then throw an execption once a non-boolean variable is quantified?

Comment (marcogario) [inline pysmt/solvers/qelim.py]: I see. What do you think of this:

- This QE declares support only for BOOL
- The check within the QE is done only on quantified variables

This solves the issue of automatic selection (it never gets selected incorrectly), but one can force the use of this qe ```qelim(f, name="shannon")```, that will raise an error if quantified variables are not bool.

Comment (mikand) [inline pysmt/solvers/qelim.py]: OK, seems reasonable

File: pysmt/test/test_shannon_expansion.py (hunk 5 above)
Comment (marcogario) [inline pysmt/test/test_shannon_expansion.py]: yes, I meant another QE. We should be able to check that the two quantifier-free expressions are equivalent:
```
f = Exists([x,y], lira_formula)
self.assertIsValid(Iff(qelim(f, name="shannon"), qelim(f)))
```

Comment (marcogario) [inline pysmt/test/test_shannon_expansion.py]: A special test was added for the mixed case.

**Pull request merged**